### DISCUSSION TITLE
[RTKQ] When using selectFromResult, return the usual properties by default as well

### DISCUSSION THREAD

**Discussion by awdyson** [reduxjs/redux-toolkit#1587] [Ideas]
`selectFromResult` is a really convenient utility, but it can grow a bit out of hand space-wise.
```
const { nestedData, isLoading } = useSomething(
  { id: 123 },
  {
    selectFromResult: ({ data, ...rest }) => ({
      nestedData: data.some.annoyingly.deep.data,
      ...rest, // I end up including this in almost every call
    })
  },
);
```

Assuming it wouldn't cause a significant performance impact, it'd be nice to just include the usual return properties, minus data. So the above could be rewritten like this (showing an ideal space-saving example):
```
// isLoading still available
const { nestedData, isLoading } = useSomething(
  { id: 123 },
  { selectFromResult: ({ data }) => ({ nestedData: data.some.annoyingly.deep.data }) },
);
```

**Comment by Shrugsy** [comment]
Realistically it *will* have a performance impact. Using [`selectFromResult`](https://redux-toolkit.js.org/rtk-query/usage/queries#selecting-data-from-a-query-result) has two main benefits:
- Adding/deriving/transforming properties returned by the hook (like getting deep nested data in your example)
- Optimising the hook to only cause a re-render when the 'selected' properties change

For your case, to optimise renders, you should *not* spread the rest of the properties, but only pass through `isLoading` from that if it is the only thing you are using. By passing through the rest, you are allowing it to re-render when any of the other properties change which you aren't using (e.g. `isFetching`, or other parts of the returned data that you aren't reading)

```js
const { nestedData, isLoading } = useSomething(
  { id: 123 },
  {
    selectFromResult: ({ data, ...rest }) => ({
      nestedData: data.some.annoyingly.deep.data,
      isLoading: rest.isLoading
    })
  },
);
```

If we change it to *always* return all data by default, then we would also need a way to opt back out of the behaviour to allow for the second dot point above. I imagine that most users would *not* want to be returning all data and would only want to return the pieces they are using, so I don't see this as being feasible.

**Reply by awdyson** [reply]
Ah nice, `selectFromResult`'s benefit is grander in scope that I realized.
Appreciate the detailed breakdown.
